### ISSUE TITLE
Can't type parameterise ProductHints for use with pureconfig

### ISSUE THREAD

**Issue by ryan1894** [pureconfig/pureconfig#468]
After running into the issue described [here](https://github.com/pureconfig/pureconfig/issues/313)

I have found that using a type parameter for a product hint:

i.e. both 
`implicit val productHint = ProductHint[T](ConfigFieldMapping(CamelCase, KebabCase).withOverrides("s3" -> "s3"))`
and
`implicit val productHint = ProductHint[SomeCaseClass](ConfigFieldMapping(CamelCase, KebabCase).withOverrides("s3" -> "s3"))`

don't apply when doing 
`pureconfig.loadConfigFromFilesOrThrow[T](Seq(Paths.get(path)))`
or `ConfigFactory.parseString[T]` or any of the other similar permutations.

but do apply when doing `pureconfig.loadConfigFromFilesOrThrow[SomeCaseClass](Seq(Paths.get(path)))`

Is there a workaround?

**Comment by leifwickland** [comment]
I'd expect that the second approach using the specific class would work if
the hint were in the scope of where the ConfigReader for SomeCaseClass
which is constructed, which is usually the call to `
loadConfigFromFilesOrThrow`. If that's not working, please consider
submitting a sample project which demonstrates the problem.

**Comment by ruippeixotog** [comment]
The product hint needs to be in scope on the place the `ConfigReader` is _created or derived_, not on the place it is used.

Typically (i.e. when a concrete class is used) the derivation would happen in your `loadConfigFromFilesOrThrow` call, but since you are using a type parameter (and there is no way to create a `ConfigReader` for an arbitrary `T`) my guess is that you're creating the `ConfigReader` somewhere else and passing it implicitly or explicitly to the class where `loadConfigFromFilesOrThrow` is called.

As @leifwickland suggested, a more complete example (preferably a runnable one) would be useful to debug your problem.

**Comment by ryan1894** [comment]
Thanks @ruippeixotog, you were correct in thinking that I was passing an implicit `ConfigReader[T]`.

Originally I was using a pattern similar to:

`case class ConfigLoader[T: ClassTag](implicit reader: ConfigReader[T], productHint: ProductHint[T]) { ... }`
and
`object Config extends ConfigLoader[SomeCaseClass](BaseConstants.RuntimeArguments.JobProfile)(ClassTag(classOf[withProductHint.JobProfileModel]), ConfigFieldMapping(CamelCase, KebabCase).withOverrides("s3" -> "s3")) { ... }`

But your comment about needing the implicit when the *creating* the `ConfigReader[T]`, I have now changed it to a pattern similar to:

```
import pureconfig.generics.auto._

object asdf {
  implicit val productHintA = ProductHint[SomeCaseClass](ConfigFieldMapping(CamelCase, KebabCase).withOverrides("s3" -> "s3"))
  lazy val Config = new ConfigLoader[SomeCaseClass]
}
```

I now realise that I needed to make sure the implicits were available when the ConfigReader is derived/created.

Thanks

Also for anybody else who might be reading this: when using nested case classes, it seems that the ProductHint should be tagged with the child class, e.g.

```
case class B(a: String)
case class A(b: B)

trait asdf {
    implicit val productHInt: ProductHint[B] = ...
    val config = new ConfigLoader[A]...
}

```
